test/unit/org/apache/cassandra/index/sai/cql/BM25Test.java

```diff
@@ -507,4 +518,37 @@ public void testQueryEmptyTable()
         var result = execute("SELECT k FROM %s ORDER BY v BM25 OF 'test' LIMIT 1");
         assertThat(result).hasSize(0);
     }
+
+    @Test
+    public void testBM25RaceConditionConcurrentQueriesInInvertedIndexSearcher() throws Throwable
+    {
+        createTable("CREATE TABLE %s (pk int, v text, PRIMARY KEY (pk))");
+        analyzeIndex();
+
+        // Create 3 docs that have the same BM25 score and will be our top docs
+        execute("INSERT INTO %s (pk, v) VALUES (1, 'apple apple apple')");
+        execute("INSERT INTO %s (pk, v) VALUES (2, 'apple apple apple')");
+        execute("INSERT INTO %s (pk, v) VALUES (3, 'apple apple apple')");
+
+        // Now insert a lot of docs that will hit the query, but will be lower in frequency and therefore in score
+        for (int i = 4; i < 1000; i++)
+            execute("INSERT INTO %s (pk, v) VALUES (?, 'apple apple')", i);
+
+        // Bug only present in sstable
+        flush();
+
+        // Trigger many concurrent queries
+        final ExecutorService executor = Executors.newFixedThreadPool(10);
+        String select = "SELECT pk FROM %s ORDER BY v BM25 OF 'apple' LIMIT 3";
+        var futures = new ArrayList<Future<UntypedResultSet>>();
+        for (int i = 0; i < 100; i++)
```

Comment:
I've run the test a few times without the fix, and there's around a 50% chance of reproducing the bug. We could increase this iteration count to maybe 1000 to make sure it's very likely to hit it.

Reply:
Sure, I'll do that.

Reply:
Looks like even with 1k queries, it's not perfect. I increased the number of rows too since that'll make each query take a bit longer to process and spend more time in the loop to read doc frequencies.